src/TheSpace/HarbergerMarket.sol

```diff
@@ -0,0 +1,200 @@
+    asset = new Asset(assetName_, assetSymbol_, address(this));
+
+    // initialize currency contract
+    currency = ERC20(currencyAddress_);
+
+    // TBD: shall we allow changes of tax rate?
+    taxRate = taxRate_;
+  }
+
+  // /**
+  //  * @dev Set address for attached currency contract.
+  //  *
+  //  */
+  // function setCurrencyAddress(address _currencyAddress) external {
+  //   require(_currencyAddress != address(0), "Address cannot be null");
+
+  //   currency = ERC20(_currencyAddress);
+  // }
+
+  /**
+   * @dev Set the current price of an Harberger property with token id.
+   *
+   * Emits a {Price} event.
+   */
+  function setPrice(uint64 _tokenId, uint256 _price) external {
+    require(asset.ownerOf(_tokenId) == msg.sender, "Sender does not own asset");
+
+    // update asset price
+    assetPrice[_tokenId] = _price;
+
+    // update market price
+    marketCap += _price;
+
+    // emit events
+    emit Price(_tokenId, _price);
+  }
+
+  /**
+   * @dev Returns the current price of an Harberger property with token id.
+   */
+  function getPrice(uint64 tokenId) external view returns (uint256 price) {
+    return assetPrice[tokenId];
+  }
+
+  /**
+   * @dev Purchase property with bid higher than current price. Clear tax for owner before transfer.
+   * TODO: check security implications
+   */
+  function bid(uint64 tokenId, uint256 price) external {
+    // TODO: mint token if not already exists, assign resource if needed
+
+    require(price >= this.getPrice(tokenId), "Price too low");
+
+    // collect tax
+    this.collectTax(asset.ownerOf(tokenId));
+
+    // TODO: handle default scenario
+    // transfer currency
+    currency.transferFrom(msg.sender, asset.ownerOf(tokenId), price);
+
+    // transfer asset
+    asset.safeTransferByMarket(asset.ownerOf(tokenId), msg.sender, tokenId);
+  }
+
+  /**
+   * @dev Collect outstanding property tax for a given address, put property on tax sale if obligation not met.
+   *
+   * Emits a {Tax} event and a {Price} event (when properties are put on tax sale).
+   */
+  function collectTax(address taxpayer) external {
+    uint256 totalAssetValue = _getAssetValue(taxpayer);
+    if (totalAssetValue > 0) {
+      // address(this).balance
+      currency.transferFrom(
+        taxpayer,
+        address(this),
+        (totalAssetValue * taxRate) / 100
+      );
+
+      // TODO: default asset if owner does not have enough
+      taxTime[taxpayer] = block.timestamp;
+    }
+  }
+
+  /**
+   * @dev Collect all outstanding property tax, put property on tax sale if obligation not met.
+   *
+   * Emits {Tax} events and {Price} events (when properties are put on tax sale).
+   */
+  function collectTaxForAll() external {
+    // TODO: collect tax for all users with old enough tax timestamp, and reward caller
+  }
+
+  /**
+   * @dev Payout all dividends from current balance.
+   *
+   * Emits {Dividend} events.
+   */
+  function distributeDividendForAll() external {
```

Comment:
can we separate into different calls? can refer to the liquidity pool dividend issue.